### PR TITLE
libtensorflow

### PR DESCRIPTION
**Pull request by njzjz**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your
      recipe (see
      [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos)
      for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

it's a try

### CODE REVIEW

**Comment by njzjz** [comment]
@conda-forge/help-c-cpp 
This is the C++ library of the Tensorflow. Please review.

Also cc [TensorFlow recipe](https://github.com/conda-forge/tensorflow-feedstock)'s maintainers: @danielfrg @dhirschfeld @dougalsutherland @farhantejani @ghego @gilbertfrancois @hajapy @jjhelmus @waitingkuo

Thank you.

**Comment by hajapy** [comment]
Wow this looks great, thanks for the hard work!

How hard would it be to also pull out the c api libtensorflow_c.so? Would we prefer this as a single recipe called libtensorflow that had multiple outputs or as separate recipes with their own feedstocks? 

My inclination, were build times not a concern, would be a single recipe with multiple outputs - it would be easier to keep these builds in sync.

**Comment by njzjz** [comment]
Do you mean something like [the recipe for blas](https://github.com/conda-forge/blas-feedstock/blob/master/recipe/meta.yaml)?

**Comment by njzjz** [comment]
> Wow this looks great, thanks for the hard work!
> 
> How hard would it be to also pull out the c api libtensorflow_c.so? Would we prefer this as a single recipe called libtensorflow that had multiple outputs or as separate recipes with their own feedstocks?
> 
> My inclination, were build times not a concern, would be a single recipe with multiple outputs - it would be easier to keep these builds in sync.

I have updated the recipe to build libtensorflow.so and libtensorflow_cc.so in a single recipe. If the build time is enough, I think we can even build C, C++, and Python version at the same recipe.

**Comment by njzjz** [comment]
I‘m glad to see all checks have passed! Now this recipe outputs two package: C API library `libtensorflow` and C++ API library `libtensorflow_cc`.

**Comment by njzjz** [comment]
@conda-forge/staged-recipes  anyone to review this recipe?

File: recipes/libtensorflow/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+package:
+  name: libtensorflow-split
+  version: {{ version }}
+
+source:
+  url: https://github.com/tensorflow/tensorflow/archive/v{{ version }}.tar.gz  # [linux]
+  sha256: 7cd19978e6bc7edc2c847bce19f95515a742b34ea5e28e4389dade35348f58ed  # [linux]
+
+  patches:
+    - patches/0001-add-rt-to-linkopts.patch  # [linux]
```
Comment (jjhelmus) [inline recipes/libtensorflow/meta.yaml]: Only this first patch is needed since the Python tests are not included nor run during the build.

File: recipes/libtensorflow/cp_libtensorflow_cc.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+mkdir -p ${PREFIX}/include/tensorflow
+cp -r bazel-genfiles/* ${PREFIX}/include/
+cp -r tensorflow/cc ${PREFIX}/include/tensorflow
+cp -r tensorflow/core ${PREFIX}/include/tensorflow
+cp -r third_party ${PREFIX}/include
+
+# link eigen
+for file in $(ls ${PREFIX}/include/eigen3)
+do
+	ln -s ${PREFIX}/include/eigen3/${file} ${PREFIX}/include/${file}
```
Comment (isuruf) [inline recipes/libtensorflow/cp_libtensorflow_cc.sh]: This shouldn't be done in this package. Why is this needed in the first place?

Comment (jjhelmus) [inline recipes/libtensorflow/cp_libtensorflow_cc.sh]: This seem a bit odd, why does the `libtensorflow_cc` package have symlinks to include files in the eigen package?

File: recipes/libtensorflow/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        - test -f $PREFIX/lib/libtensorflow.so  # [not win]
+
+  - name: libtensorflow_cc
+    script: cp_libtensorflow_cc.sh  # [unix]
+    requirements:
+      host:
+        - libprotobuf
+        - eigen
+        - nsync
+        - abseil-cpp
```
Comment (jjhelmus) [inline recipes/libtensorflow/meta.yaml]: Are these host requirements needed?  Other than eigen they do not seem used in the install script.

File: recipes/libtensorflow/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    test:
+      commands:
+        - test -f $PREFIX/lib/libtensorflow_cc.so  # [not win]
+
+about:
+  home: http://tensorflow.org/
+  license: Apache 2.0
+  license_file: LICENSE
+  license_family: Apache
+  summary: TensorFlow is an open source machine learning framework for everyone.
```
Comment (jjhelmus) [inline recipes/libtensorflow/meta.yaml]: It would be good to add text here explaining that these are the C++ components.

**Comment by isuruf** [comment]
Is this library built when the python package is built as well? If so, this can be an `output`.

**Comment by hajapy** [comment]
I think the python package is a different build target but if there was a way to consolidate all these outputs to the tensorflow feedstock it would be great as again it would simplify the maintenance and keep the recipes in sync.

**Comment by njzjz** [comment]
> I think the python package is a different build target but if there was a way to consolidate all these outputs to the tensorflow feedstock it would be great as again it would simplify the maintenance and keep the recipes in sync.

I tried conda-forge/tensorflow-feedstock#79 but it failed. I am not sure the reason.

Anyway, in this recipe I have make changes according to the above reviews.

File: recipes/libtensorflow/cp_libtensorflow_cc.sh (hunk 2 above)
Comment (njzjz) [reply recipes/libtensorflow/cp_libtensorflow_cc.sh]: It's required to add `eigen3` to include directories when using C++ API. Indeed, this step can be done when compiling with C++ API, so I have removed it.

File: recipes/libtensorflow/meta.yaml (hunk 3 above)
Comment (njzjz) [reply recipes/libtensorflow/meta.yaml]: I have removed all host requirements.

**Comment by njzjz** [comment]
By the way, is the CI time enough if we build them in one recipe? `linux_python3.7` has consumed 356 min in conda-forge/tensorflow-feedstock#79 although it fails.

**Comment by njzjz** [comment]
> I think the python package is a different build target but if there was a way to consolidate all these outputs to the tensorflow feedstock it would be great as again it would simplify the maintenance and keep the recipes in sync.

I fixed the issue in conda-forge/tensorflow-feedstock#79 but the CI time is not enough. Is it better to still use this recipe? @isuruf @hajapy
